crates/bevy_ecs/src/storage/sorted_small_vec.rs

```diff
@@ -0,0 +1,259 @@
+    }
+
+    pub(crate) fn ones(&self) -> impl Iterator<Item = usize> + '_ {
+        self.0.iter().copied()
+    }
+
+    /// Insert the value if it's not already present in the vector.
+    /// Maintains a sorted order.
+    pub(crate) fn insert(&mut self, index: usize) {
+        match self.0.binary_search(&index) {
+            // element already present in the vector
+            Ok(_) => {}
+            Err(pos) => {
+                self.0.insert(pos, index);
+            }
+        }
+    }
+
+    /// Removes a value if it's present in the vector
+    pub(crate) fn remove(&mut self, index: usize) {
+        if let Ok(pos) = self.0.binary_search(&index) {
+            self.0.remove(pos);
+        }
+   }
+
+    /// Returns true if the vector contains the value.
+    pub(crate) fn contains(&self, index: usize) -> bool {
+        self.0.binary_search(&index).is_ok()
+    }
+
+    /// Returns true if the vector is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Empties the contents of the vector
+    pub(crate) fn clear(&mut self) {
+        self.0.clear();
+    }
+
+    /// Returns the number of elements in the vector.
+    pub(crate) fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Adds all the elements from `other` into this vector. (skipping duplicates)
+    pub(crate) fn union_with(&mut self, other: &Self) {
+        let mut i = 0;
+        let mut j = 0;
+        while i < self.len() && j < other.len() {
+            match self.0[i].cmp(&other.0[j]) {
+                Ordering::Less => i += 1,
+                Ordering::Greater => {
+                    self.0.insert(i, other.0[j]);
+                    j += 1;
+                }
+                Ordering::Equal => {
+                    i += 1;
+                    j += 1;
+                }
+            }
+        }
+        while j < other.len() {
+            self.0.push(other.0[j]);
+            j += 1;
+        }
+    }
+
+    /// Returns the elements that are in both `self` and `other`.
+    pub(crate) fn intersection<'a>(&'a self, other: &'a Self) -> Intersection<'a, N> {
+        Intersection {
+            this: self,
+            other,
+            i: 0,
+            j: 0,
+        }
+    }
+
+    /// Return the elements that are in `self` but not in `other`.
+    pub(crate) fn difference<'a>(&'a self, other: &'a Self) -> Difference<'a, N> {
+        Difference {
+            this: self,
+            other,
+            i: 0,
+            j: 0,
+        }
+    }
+
+    /// Returns true if the two vectors have no common elements.
+    pub(crate) fn is_disjoint(&self, other: &Self) -> bool {
+        self.intersection(other).next().is_none()
+    }
+
+    /// Returns true if all the elements in `self` are also in `other`.
+    pub(crate) fn is_subset(&self, other: &Self) -> bool {
+        self.difference(other).next().is_none()
+    }
+}
+
+impl<const N: usize> Extend<usize> for SortedSmallVec<N> {
```

Comment:
This implementation is wrong if `other` is unsorted. So we should either make this just a loop of inserts or add a method instead of a trait impl that makes it clear the behavior is undefined if you don't call it with a sorted input.

Reply:
I think you can handle unsorted input and still be reasonably efficient with just:
```rust
self.0.extend(other);
self.0.sort();
self.0.dedup();
```